addons/mrp/models/product.py

```diff
@@ -429,3 +429,39 @@ def action_archive(self):
 
     def _get_backend_root_menu_ids(self):
         return super()._get_backend_root_menu_ids() + [self.env.ref('mrp.menu_mrp_root').id]
+
+    def _update_uom(self, to_uom_id):
+        error_message = ('As other units of measure (ex : %(problem_uom)s) '
+        'than %(uom)s have already been used for this product, the change of unit of measure can not be done.'
+        'If you want to change it, please archive the product and create a new one.')
+        for uom, product, bom_ids in self.env['mrp.bom']._read_group(
+            [('product_tmpl_id', 'in', self.product_tmpl_id.ids)],
+            ['product_uom_id', 'product_id'],
+            ['id:array_agg'],
+        ):
+            boms = self.env['mrp.bom'].browse(bom_ids)
+            if product.product_tmpl_id.uom_id != uom:
+                raise UserError(_(error_message, problem_uom=uom.name, uom=product.product_tmpl_id.uom_id.name))
+            boms.product_uom_id = to_uom_id
```

Comment:
Shouldn't we delay the BOMs correction to the end of the function until we make sure that BOM lines and MOs are also fine? So that we don't end up in an inconsistent state (only BOM units are updated but not BOM lines and MOs for example). Same applies for BOM lines. Or does the whole transaction abort if an error is thrown?

Reply:
When there is an error the whole transaction is aborted. There is an implicit
BEGIN;
...
COMMIT;
at each transaction